recipes/lora_ppo_single_device.py

```diff
@@ -0,0 +1,813 @@
+        self.ppo_epochs = cfg.ppo_epochs
+        self.ppo_batch_size = cfg.ppo_batch_size
+        self._gradient_accumulation_steps = cfg.gradient_accumulation_steps
+        if self.ppo_batch_size % self._gradient_accumulation_steps != 0:
+            raise ValueError(
+                f"ppo_batch_size ({self.ppo_batch_size})  must be "
+                f"exactly divisible by gradient_accumulation_steps ({cfg._gradient_accumulation_steps})."
+            )
+        self.ppo_backward_batch_size = (
+            cfg.ppo_batch_size // self._gradient_accumulation_steps
+        )
+
+        # trajectory generation args
+        self.temperature = cfg.temperature + 1e-7  # avoid underflow for 0 temperature
+        self.top_k = cfg.top_k
+        self.forward_batch_size = cfg.forward_batch_size
+        self.max_generated_tokens = cfg.max_generated_tokens
+
+        # reward masking args
+        self.truncate_after_tokens = cfg.truncate_after_tokens
+        self.penalise_no_eos = cfg.penalise_no_eos
+        self.reward_penalty = cfg.reward_penalty
+
+        self.total_epochs = self.num_steps // self.batch_size
+        self._gradient_accumulation_steps = cfg.gradient_accumulation_steps
+        # one "step" is an update over a batch of trajectories
+        self._steps_per_epoch = len(self._dataloader) // self.batch_size
+        self.global_step = self.epochs_run * self._steps_per_epoch
+
+        # setup adaptive KL controller
+        self.kl_controller = config.instantiate(cfg.kl_controller)
+
+        self._profiler_enabled = cfg.profiler.enabled
+        self._profiler = config.instantiate(cfg.profiler)
+
+    def _setup_model(
+        self,
+        cfg_model: DictConfig,
+        cfg_reward_model: DictConfig,
+        enable_activation_checkpointing: bool,
+        model_state_dict: Dict[str, Any],
+        reward_model_state_dict: Dict[str, Any],
+        model_lora_weights_state_dict: Optional[Dict[str, Any]] = None,
+        value_head_state_dict: Optional[Dict[str, Any]] = None,
+        initialise_value_head_from_reward_model: bool = False,
+    ) -> nn.Module:
+
+        with utils.set_default_dtype(self._dtype), self._device:
+            model = config.instantiate(cfg_model)
+            reward_model = config.instantiate(cfg_reward_model)
+
+        self._lora_rank = cfg_model.lora_rank
+        self._lora_alpha = cfg_model.lora_alpha
+        self.adapter_params = get_adapter_params(model)
+        set_trainable_params(model, self.adapter_params)
+
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                model, auto_wrap_policy={TransformerLMWithValueHead}
+            )
+
+        # add placeholder value head key to base model state dict keys
+        # to ensure wrapped model is validated correctly
+        model_state_dict_keys = list(model_state_dict.keys())
+        model_state_dict_keys += ["value_head.weight", "value_head.bias"]
```

Comment:
I've been reworking this whole bit since the lora finetuning recipe does things a little differently - I grabbed my lora validation from the lora dpo recipe. agree w your points, the updated implementation should be much cleaner